frontend/ui/elements/menu_keyboard_layout.lua

```diff
@@ -128,10 +132,17 @@ for k, __ in FFIUtil.orderedPairs(VirtualKeyboard.lang_to_keyboard_layout) do
             G_reader_settings:saveSetting("keyboard_layouts", keyboard_layouts)
         end,
         hold_callback = function(touchmenu_instance)
-            G_reader_settings:saveSetting("keyboard_layout_default", k)
+            G_reader_settings:saveSetting("keyboard_layout_default", lang)
             if touchmenu_instance then touchmenu_instance:updateItems() end
         end,
     })
+    local ok, keyboard = pcall(require, "ui/data/keyboardlayouts/" .. keyboard_layout)
+    if ok and keyboard.genMenuItems ~= nil then
+        table.insert(sub_item_table[4].sub_item_table, {
+            text = Language:getLanguageName(lang),
+            sub_item_table = keyboard:genMenuItems(),
+        })
+    end
```

Comment:
We didn't do this before. There isn't a cached copy of all of the loaded keyboards anywhere (each time we switch we re-do a `require`) so I thought making this more generic would be good -- but you're right it's more expensive for all of the keyboards that don't have settings.

I'll hardcode the Japanese keyboard here, but maybe we should just load and cache handles to all of the keyboards so we can just re-use those rather than reloading them each time?

Reply:
> There isn't a cached copy of all of the loaded keyboards anywhere (each time we switch we re-do a require)

require is cached, for the record.

Reply:
> but maybe we should just load and cache handles to all of the keyboards so we can just re-use those rather than reloading them each time?

Not "all of the keyboard", that's just useless memory wasted. But if you want to cache the ones that were really instantiated once because needed/used, why not.

Reply:
> require is cached, for the record.

Ah, TIL.

> But if you want to cache the ones that were really instantiated once because needed/used, why not.

Since this code is only executed once (since `require` is cached), there's probably no need to deal with any caching.

Reply:
> Ah, TIL.

Hence why we have to do stuff like this, otherwise you're just reusing it:

https://github.com/koreader/koreader/blob/7baeebc2d41244020cc7d8cb5249a5a64d0e401c/spec/unit/commonrequire.lua#L68-L86

There's one minor caveat there, in that it's cached in a global variable (see above, `package.loaded`). A `local` reference is almost always more performant for lookups, hence the `local var = require …` pattern before you start using to limit the more expensive lookup to only one occurrence.

Reply:
I also wasn't sure if `pcall` is needed here -- we don't `pcall` the keyboard loading in `VirtualKeyboard:init` but I guess crashing during menu generation is worse? Maybe I should use `dbg.dassert` here?

Reply:
Only if you're planning to make it super easy to run random custom keyboard layouts, otherwise there doesn't seem to be much point to it.

Reply:
Yeah in that case I'll drop the `pcall` :smile_cat:.